static-site/app/blog/utils.ts

```diff
@@ -22,6 +22,28 @@ export function getBlogPosts(): BlogPost[] {
 
   const postsDirectory = path.join(__dirname, "..", "..", "content", "blog");
 
+  // if this code is running in production, it won't have access to
+  // the postsDirectory, because it doesn't exist there — so check to
+  // see if we can `stat()` that path, and if we get an error, if it's
+  // a "file not found" error, just return an empty list; otherwise,
+  // something else weird is going on, and the best we can do is
+  // re-throw the error
+  try {
+    fs.statSync(postsDirectory);
+  }
+  catch (err: unknown) {
+    if (
+      typeof err === "object" &&
+      err !== null &&
+        Object.prototype.hasOwnProperty.call(err, "code") &&
+      err["code"] === "ENOENT"
+    ) {
```

Comment:
_(non-blocking)_

Simpler and should work equally well:

```suggestion
    if (
      err instanceof Error &&
      'code' in err &&
      err["code"] === "ENOENT"
    ) {
```

Reply:
yeah, thanks, applied that change, that's better than the pretzel mine was twisted into by the time the linter was happy.

Reply:
That would work with `err: any`, but I think it's good practice to use `unknown` instead, even if it requires a couple more type guards.

Reply:
Ah, I see. Thanks!

More reading:
- https://www.typescriptlang.org/docs/handbook/2/functions.html#unknown
- https://devblogs.microsoft.com/typescript/announcing-typescript-3-0-rc-2/#the-unknown-type

Reply:
Ok, another question, if you'll humor me: why doesn't TS know that Node's `Error` class has a [`code` property](https://nodejs.org/api/errors.html#errorcode)? i.e. why is `"code" in err` necessary?

Reply:
Yeah, i'd like to know this too. Even if you subclass `Error` and `delete this.code` you can still access `err.code` directly, just like you can do `({}).whatever`.

Reply:
> Ok, another question, if you'll humor me: why doesn't TS know that Node's `Error` class has a [`code` property](https://nodejs.org/api/errors.html#errorcode)? i.e. why is `"code" in err` necessary?

After a bit of local testing, I think it probably isn't -- this code seems to be fine:

```
    if (
      err instanceof Error &&
      err["code"] === "ENOENT"
    ) {
```

…and this version is throwing a typing error in my editor but seems to lint okay? :shrug:

```
    if (
      err instanceof Error &&
      err.code === "ENOENT"
    ) {
```

Reply:
> why doesn't TS know that Node's `Error` class has a [`code` property](https://nodejs.org/api/errors.html#errorcode)?

TypeScript limitations are showing here 😅

The [built-in type definition](https://github.com/microsoft/TypeScript/blob/71716a2868c87248af5020e33a84a2178d41a2d6/src/lib/es2022.error.d.ts#L7-L9) for `Error` doesn't contain the `code` property because it is reflective of [JavaScript's Error class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#instance_properties), not Node.js's Error class. The latter is [defined by `@types/node`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/ab4715a8aead1731ced56619128c31e57bcc682d/types/node/v16/globals.d.ts#L167), however we can't use `err instanceof NodeJS.ErrnoException` [because it's not available at run time](https://github.com/threema-ch/threema-desktop/blob/3e1f4ea5a04ee59d745fb20a1a96e36051a51c05/src/common/node/utils.ts#L10-L13).

Reply:
> [`err["code"]`] seems to be fine

This is allowed because we [bypass `noImplicitAny`](https://github.com/nextstrain/nextstrain.org/blob/7d4d6a8c46a60b549a09201de38d5fcdf447bea4/static-site/tsconfig.json#L32). +1 for [removing the bypass](https://github.com/nextstrain/nextstrain.org/issues/879).

> [`err.code`] is throwing a typing error in my editor but seems to lint okay

It errors when running `npm run type-check`: `error TS2339: Property 'code' does not exist on type 'Error'.`

Reply:
> Even if you subclass `Error` and `delete this.code` you can still access `err.code` directly, just like you can do `({}).whatever`.

You can do many things in JavaScript that are arguably prone to error and TypeScript helps reduce the potential for error. I think for this particular case, the guardrails set up by TypeScript are not very useful, as checking `err.code === "ENOENT"` already handles the case where it is `undefined`, so something like this is sufficient:

```ts
  catch (err: any) {
    if (
      err?.code === "ENOENT"
    )
```

but it's nice to get into the habit of using `unknown` when possible.

Reply:
Thanks for helping me learn a bit more about the nuances TypeScript! 

My high-level takeaway is that I still dislike (though also still find value in) type checking that is bolted onto languages rather than an intrinsic part of them.